relayer/src/link.rs

```diff
@@ -841,30 +1246,238 @@ impl RelayPath {
+                }
+            }
+            // Update the whole batch, keeping only the relevant ones
+            odata.batch = retain_batch;
+        }
+
+        if timed_out.is_empty() {
+            // Nothing timed out in the meantime
+            Ok(())
+        } else {
+            // Schedule new operational data targeting the source chain
+            for (&_pos, batch) in timed_out.iter() {
+                let mut new_od =
+                    OperationalData::new(dst_current_height, OperationalDataTarget::Source);
+                new_od.batch = batch.clone();
+
+                info!(
+                    "[{}] Scheduling from batch of size {}",
+                    self,
+                    new_od.batch.len()
+                );
+
+                self.schedule_operational_data(new_od)?;
+            }
+
+            self.dst_operational_data = all_dst_odata;
+            // Possibly some op. data became empty (if no events were kept).
+            // Retain only the non-empty ones.
+            self.dst_operational_data
+                .retain(|odata| !odata.batch.is_empty());
+            Ok(())
+        }
+    }
+
+    /// Adds a new operational data item for this relaying path to process later.
+    /// If the relaying path has non-zero packet delays, this method also updates the client on the
+    /// target chain with the appropriate headers.
+    fn schedule_operational_data(&mut self, mut od: OperationalData) -> Result<(), LinkError> {
+        if od.batch.is_empty() {
+            warn!(
+                "[{}] Ignoring operational data for {} because it has no messages",
+                self, od.target
+            );
+            return Ok(());
+        }
+        info!(
+            "[{}] Scheduling op. data with {} msg(s) for target chain ({}) (height {})",
+            self,
+            od.batch.len(),
+            od.target,
+            od.proofs_height,
+        );
+
+        // Update clients ahead of scheduling the operational data, if the delays are non-zero.
+        if !self.zero_delay() {
+            let target_height = od.proofs_height.increment();
+            match od.target {
+                OperationalDataTarget::Source => self.update_client_src(target_height)?,
+                OperationalDataTarget::Destination => self.update_client_dst(target_height)?,
+            };
+        }
+
+        od.scheduled_time = Instant::now();
+
+        match od.target {
+            OperationalDataTarget::Source => self.src_operational_data.push(od),
+            OperationalDataTarget::Destination => self.dst_operational_data.push(od),
+        };
+
+        Ok(())
+    }
+
+    /// Pulls out the next operational data that has fulfilled the predefined delay period and can
+    /// now be processed. Does not block: if no OD fulfilled the delay period (or none is
+    /// scheduled), returns immediately with `None`.
+    fn try_fetch_scheduled_operational_data(&mut self) -> Option<OperationalData> {
```

Comment:
I think should return the data for both src and dst and send both in `step()`. Otherwise, if there are often things to schedule from `src`, the `dst` may be delayed significantly.

Reply:
fixed now